### ISSUE TITLE
Assertion lowerValue<=upperValue failed 

### ISSUE THREAD

**Issue by benoit9126** [coin-or/CyLP#76]
Hi,

Here is my problem with `Cbc` and `CyLP`. I model a integer linear program using `cvxpy` and then send it to `Cbc` (so via `CyLP`) and I got an annoying assertion error in the code of `Cbc`:

```console
$ python3 essai.py # File which reads the data of the problem and solve it through cvxpy (and cylp)
Welcome to the CBC MILP Solver 
Version: 2.9.9 
Build Date: Aug 21 2017 

command line - ICbcModel -solve -quit (default strategy 1)
Continuous objective value is 18.623 - 0.01 seconds
Cgl0006I 59 SOS (683 members out of 946) with 532 overlaps - too much overlap or too many others
Cgl0009I 728 elements changed
Cgl0003I 44 fixed, 0 tightened bounds, 2 strengthened rows, 0 substitutions
Cgl0009I 539 elements changed
Cgl0004I processed model has 321 rows, 777 columns (777 integer (777 of which binary)) and 15166 elements
Cbc0038I Initial state - 2 integers unsatisfied sum - 0.5
Cbc0038I Pass   1: suminf.    0.50000 (2) obj. 18.623 iterations 3
Cbc0038I Pass   2: suminf.    0.66667 (2) obj. 19.0397 iterations 2
Cbc0038I Pass   3: suminf.    5.50000 (13) obj. 35.3779 iterations 147
Cbc0038I Pass   4: suminf.    3.83333 (9) obj. 34.5446 iterations 24
Cbc0038I Pass   5: suminf.    0.66667 (2) obj. 35.6279 iterations 13
Cbc0038I Pass   6: suminf.    0.50000 (2) obj. 35.2112 iterations 7
Cbc0038I Pass   7: suminf.    6.91667 (16) obj. 43.5571 iterations 112
Cbc0038I Pass   8: suminf.    2.66667 (7) obj. 41.9321 iterations 42
Cbc0038I Pass   9: suminf.    1.66667 (5) obj. 41.9321 iterations 2
Cbc0038I Pass  10: suminf.    0.66667 (2) obj. 41.4321 iterations 5
Cbc0038I Pass  11: suminf.    0.50000 (2) obj. 41.0155 iterations 2
Cbc0038I Pass  12: suminf.    6.00000 (13) obj. 42.0363 iterations 109
Cbc0038I Pass  13: suminf.    4.00000 (9) obj. 41.5363 iterations 30
Cbc0038I Pass  14: suminf.    3.50000 (8) obj. 41.2863 iterations 1
Cbc0038I Pass  15: suminf.    3.00000 (7) obj. 41.0363 iterations 1
Cbc0038I Solution found of 40.0363
Cbc0038I Before mini branch and bound, 593 integers at bound fixed and 0 continuous
Cbc0038I Full problem 321 rows 777 columns, reduced to 78 rows 90 columns
Cbc0038I Mini branch and bound improved solution from 40.0363 to 19.5288 (0.05 seconds)
Cbc0038I Round again with cutoff of 19.4382
Cbc0038I Reduced cost fixing fixed 20 variables on major pass 2
Cbc0038I Pass  16: suminf.    0.50000 (2) obj. 18.623 iterations 0
Cbc0038I Pass  17: suminf.    0.66667 (2) obj. 19.0397 iterations 2
Cbc0038I Pass  18: suminf.    1.77786 (7) obj. 19.4382 iterations 45
Cbc0038I Pass  19: suminf.    1.42220 (7) obj. 19.4382 iterations 34
Cbc0038I Pass  20: suminf.    1.56744 (6) obj. 19.4382 iterations 4
Cbc0038I Pass  21: suminf.    1.56744 (6) obj. 19.4382 iterations 1
Cbc0038I Pass  22: suminf.    1.21521 (4) obj. 19.4382 iterations 13
Cbc0038I Pass  23: suminf.    1.21521 (4) obj. 19.4382 iterations 0
Cbc0038I Pass  24: suminf.    1.28479 (4) obj. 19.4382 iterations 1
Cbc0038I Pass  25: suminf.    1.08333 (4) obj. 19.3375 iterations 6
Cbc0038I Pass  26: suminf.    1.21521 (4) obj. 19.4382 iterations 6
Cbc0038I Pass  27: suminf.    1.64923 (7) obj. 19.4382 iterations 43
Cbc0038I Pass  28: suminf.    1.64923 (7) obj. 19.4382 iterations 22
Cbc0038I Pass  29: suminf.    1.23084 (4) obj. 19.4382 iterations 16
Cbc0038I Pass  30: suminf.    0.75000 (3) obj. 19.1978 iterations 11
Cbc0038I Pass  31: suminf.    0.89751 (3) obj. 19.4382 iterations 11
Cbc0038I Pass  32: suminf.    0.83333 (3) obj. 19.4061 iterations 10
Cbc0038I Pass  33: suminf.    0.89751 (3) obj. 19.4382 iterations 11
Cbc0038I Pass  34: suminf.    0.83333 (3) obj. 19.4061 iterations 9
Cbc0038I Pass  35: suminf.    0.89751 (3) obj. 19.4382 iterations 11
Cbc0038I Pass  36: suminf.    2.50679 (8) obj. 19.4382 iterations 46
Cbc0038I Pass  37: suminf.    1.73999 (7) obj. 19.4382 iterations 22
Cbc0038I Pass  38: suminf.    1.20970 (4) obj. 19.4382 iterations 16
Cbc0038I Pass  39: suminf.    0.75000 (3) obj. 19.2083 iterations 13
Cbc0038I Pass  40: suminf.    0.87637 (3) obj. 19.4382 iterations 11
Cbc0038I Pass  41: suminf.    0.83333 (3) obj. 19.4167 iterations 10
Cbc0038I Pass  42: suminf.    0.87637 (3) obj. 19.4382 iterations 12
Cbc0038I Pass  43: suminf.    0.83333 (3) obj. 19.4167 iterations 10
Cbc0038I Pass  44: suminf.    0.87637 (3) obj. 19.4382 iterations 14
Cbc0038I Pass  45: suminf.    0.78277 (8) obj. 19.4382 iterations 46
Cbc0038I No solution found this major pass
Cbc0038I Before mini branch and bound, 730 integers at bound fixed and 0 continuous
Cbc0038I Mini branch and bound did not improve solution (0.07 seconds)
Cbc0038I After 0.07 seconds - Feasibility pump exiting with objective of 19.5288 - took 0.04 seconds
Cbc0012I Integer solution of 19.528782 found by feasibility pump after 0 iterations and 0 nodes (0.07 seconds)
Cbc0038I Full problem 321 rows 777 columns, reduced to 0 rows 0 columns
Cbc0006I The LP relaxation is infeasible or too expensive
Cbc0031I 3 added rows had average density of 9
Cbc0013I At root node, 3 cuts changed objective from 18.62299 to 19.362782 in 2 passes
Cbc0014I Cut generator 0 (Probing) - 2 row cuts average 1.0 elements, 6 column cuts (6 active)  in 0.003 seconds - new frequency is 1
Cbc0014I Cut generator 1 (Gomory) - 2 row cuts average 11.5 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is 1
Cbc0014I Cut generator 2 (Knapsack) - 5 row cuts average 9.4 elements, 0 column cuts (0 active)  in 0.001 seconds - new frequency is 1
Cbc0014I Cut generator 3 (Clique) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 4 (MixedIntegerRounding2) - 2 row cuts average 6.0 elements, 0 column cuts (0 active)  in 0.001 seconds - new frequency is 1
Cbc0014I Cut generator 5 (FlowCover) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 6 (TwoMirCuts) - 5 row cuts average 5.4 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is 1
python3: ClpNonLinearCost.cpp:1035: void ClpNonLinearCost::checkInfeasibilities(double): Assertion `lowerValue<=upperValue' failed.
Abandon
```

When I write the same problem into a mps file and then run `Cbc` on it from command line, it works fine...

```console
$ cbc essai.mps 
Welcome to the CBC MILP Solver 
Version: 2.9.9 
Build Date: Aug 21 2017 

command line - cbc essai.mps (default strategy 1)
At line 1 NAME          ClpDefau
At line 2 ROWS
At line 521 COLUMNS
At line 18455 RHS
At line 18675 BOUNDS
At line 19622 ENDATA
Problem ClpDefau has 517 rows, 946 columns and 34782 elements
Coin0008I ClpDefau read with 0 errors
Continuous objective value is 18.623 - 0.03 seconds
Cgl0003I 44 fixed, 0 tightened bounds, 2 strengthened rows, 0 substitutions
Cgl0004I processed model has 321 rows, 777 columns (777 integer (777 of which binary)) and 15166 elements
Cbc0038I Initial state - 2 integers unsatisfied sum - 0.5
Cbc0038I Pass   1: suminf.    0.50000 (2) obj. 18.623 iterations 1
Cbc0038I Pass   2: suminf.    0.66667 (2) obj. 19.0397 iterations 2
Cbc0038I Pass   3: suminf.    5.50000 (13) obj. 35.3779 iterations 146
Cbc0038I Pass   4: suminf.    3.83333 (9) obj. 34.5446 iterations 24
Cbc0038I Pass   5: suminf.    3.33333 (8) obj. 34.7946 iterations 1
Cbc0038I Pass   6: suminf.    2.83333 (7) obj. 35.0446 iterations 1
Cbc0038I Pass   7: suminf.    0.66667 (2) obj. 35.6279 iterations 11
Cbc0038I Pass   8: suminf.    0.50000 (2) obj. 35.2112 iterations 7
Cbc0038I Pass   9: suminf.    6.58333 (15) obj. 41.1418 iterations 104
Cbc0038I Pass  10: suminf.    2.75000 (7) obj. 39.2251 iterations 38
Cbc0038I Pass  11: suminf.    1.75000 (5) obj. 39.2251 iterations 2
Cbc0038I Pass  12: suminf.    0.33333 (1) obj. 38.6835 iterations 5
Cbc0038I Pass  13: suminf.    0.25000 (1) obj. 38.4751 iterations 1
Cbc0038I Pass  14: suminf.    6.66667 (15) obj. 44.7077 iterations 109
Cbc0038I Pass  15: suminf.    2.83333 (7) obj. 42.791 iterations 30
Cbc0038I Pass  16: suminf.    1.83333 (5) obj. 42.291 iterations 2
Cbc0038I Solution found of 41.2077
Cbc0038I Before mini branch and bound, 605 integers at bound fixed and 0 continuous
Cbc0038I Full problem 321 rows 777 columns, reduced to 69 rows 84 columns
Cbc0038I Mini branch and bound improved solution from 41.2077 to 19.6389 (0.17 seconds)
Cbc0038I Round again with cutoff of 19.5373
Cbc0038I Reduced cost fixing fixed 12 variables on major pass 2
Cbc0038I Pass  17: suminf.    0.50000 (2) obj. 18.623 iterations 0
Cbc0038I Pass  18: suminf.    0.66667 (2) obj. 19.0397 iterations 2
Cbc0038I Pass  19: suminf.    2.55467 (9) obj. 19.5373 iterations 45
Cbc0038I Pass  20: suminf.    1.84018 (7) obj. 19.5373 iterations 31
Cbc0038I Pass  21: suminf.    1.25000 (5) obj. 19.4431 iterations 16
Cbc0038I Pass  22: suminf.    1.39501 (5) obj. 19.5373 iterations 7
Cbc0038I Pass  23: suminf.    1.41308 (5) obj. 19.5373 iterations 1
Cbc0038I Pass  24: suminf.    1.38411 (4) obj. 19.5373 iterations 2
Cbc0038I Pass  25: suminf.    1.25000 (4) obj. 19.4702 iterations 7
Cbc0038I Pass  26: suminf.    1.11589 (4) obj. 19.5373 iterations 6
Cbc0038I Pass  27: suminf.    1.11589 (4) obj. 19.5373 iterations 0
Cbc0038I Pass  28: suminf.    1.68162 (9) obj. 19.5373 iterations 60
Cbc0038I Pass  29: suminf.    0.75000 (3) obj. 19.3764 iterations 35
Cbc0038I Pass  30: suminf.    0.92829 (3) obj. 19.5373 iterations 8
Cbc0038I Pass  31: suminf.    0.92829 (3) obj. 19.5373 iterations 0
Cbc0038I Pass  32: suminf.    0.92829 (3) obj. 19.5373 iterations 11
Cbc0038I Pass  33: suminf.    0.92829 (3) obj. 19.5373 iterations 0
Cbc0038I Pass  34: suminf.    0.92829 (3) obj. 19.5373 iterations 2
Cbc0038I Pass  35: suminf.    1.62371 (8) obj. 19.5373 iterations 51
Cbc0038I Pass  36: suminf.    1.00000 (4) obj. 19.4664 iterations 36
Cbc0038I Pass  37: suminf.    1.14174 (4) obj. 19.5373 iterations 9
Cbc0038I Pass  38: suminf.    2.90389 (9) obj. 19.5373 iterations 52
Cbc0038I Pass  39: suminf.    1.95969 (7) obj. 19.5373 iterations 35
Cbc0038I Pass  40: suminf.    1.15844 (4) obj. 19.5373 iterations 19
Cbc0038I Pass  41: suminf.    0.75000 (3) obj. 19.3331 iterations 8
Cbc0038I Pass  42: suminf.    0.84156 (3) obj. 19.5373 iterations 7
Cbc0038I Pass  43: suminf.    0.84156 (3) obj. 19.5373 iterations 0
Cbc0038I Pass  44: suminf.    0.84156 (3) obj. 19.5373 iterations 11
Cbc0038I Pass  45: suminf.    0.84156 (3) obj. 19.5373 iterations 0
Cbc0038I Pass  46: suminf.    0.84156 (3) obj. 19.5373 iterations 2
Cbc0038I No solution found this major pass
Cbc0038I Before mini branch and bound, 718 integers at bound fixed and 0 continuous
Cbc0038I Mini branch and bound did not improve solution (0.25 seconds)
Cbc0038I After 0.25 seconds - Feasibility pump exiting with objective of 19.6389 - took 0.14 seconds
Cbc0012I Integer solution of 19.638886 found by feasibility pump after 0 iterations and 0 nodes (0.25 seconds)
Cbc0012I Integer solution of 19.528782 found by DiveCoefficient after 17 iterations and 0 nodes (0.30 seconds)
Cbc0031I 6 added rows had average density of 8.5
Cbc0013I At root node, 6 cuts changed objective from 18.62299 to 19.528782 in 5 passes
Cbc0014I Cut generator 0 (Probing) - 5 row cuts average 2.4 elements, 4 column cuts (4 active)  in 0.006 seconds - new frequency is 1
Cbc0014I Cut generator 1 (Gomory) - 12 row cuts average 9.0 elements, 0 column cuts (0 active)  in 0.001 seconds - new frequency is 1
Cbc0014I Cut generator 2 (Knapsack) - 8 row cuts average 7.6 elements, 0 column cuts (0 active)  in 0.003 seconds - new frequency is 1
Cbc0014I Cut generator 3 (Clique) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 4 (MixedIntegerRounding2) - 3 row cuts average 6.0 elements, 0 column cuts (0 active)  in 0.006 seconds - new frequency is 1
Cbc0014I Cut generator 5 (FlowCover) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.001 seconds - new frequency is -100
Cbc0014I Cut generator 6 (TwoMirCuts) - 17 row cuts average 6.9 elements, 0 column cuts (0 active)  in 0.003 seconds - new frequency is 1
Cbc0001I Search completed - best objective 19.52878185860001, took 17 iterations and 0 nodes (0.30 seconds)
Cbc0035I Maximum depth 0, 309 variables fixed on reduced cost
Cuts at root node changed objective from 18.623 to 19.5288
Probing was tried 5 times and created 9 cuts of which 0 were active after adding rounds of cuts (0.006 seconds)
Gomory was tried 5 times and created 12 cuts of which 0 were active after adding rounds of cuts (0.001 seconds)
Knapsack was tried 5 times and created 8 cuts of which 0 were active after adding rounds of cuts (0.003 seconds)
Clique was tried 5 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
MixedIntegerRounding2 was tried 5 times and created 3 cuts of which 0 were active after adding rounds of cuts (0.006 seconds)
FlowCover was tried 5 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.001 seconds)
TwoMirCuts was tried 5 times and created 17 cuts of which 0 were active after adding rounds of cuts (0.003 seconds)

Result - Optimal solution found

Objective value:                19.52878186
Enumerated nodes:               0
Total iterations:               17
Time (CPU seconds):             0.30
Time (Wallclock seconds):       0.13

Total time (CPU seconds):       0.33   (Wallclock seconds):       0.14
```

If I try to compile with `NDEBUG` flag, I got an error later which triggers an `abort();` call in `Cbc` code... Not much better than the assert...

1. Is there a way to get a proper error raised instead of a raw assert? 
2. Is this possible to check this kind of potential problems before triggering the `solve` function?
3. My objective is linear: why the code of `ClpNonLinearCost::checkInfeasibilities` (**NonLinear**) is triggered?
4. If there is no other options than to write a mps file and then call `cbc` from command line to avoid this kind of behaviour, is there a way through `CyLP` to read solution files written by the direct call of `cbc`?

Thanks.

**Comment by benoit9126** [comment]
Ok. I simplified my problem: I realized that some coefficients were very small (but not zeros). I force them to be exactly zeros and it works directly from Python. Writing these coefficients in a MPS file must have rounded them...
Sorry for the disturbance.

**Comment by kevinbishop-eaua** [comment]
Hi @benoit9126 I realize this is a very old thread, but I'm running into the same issue.

What method did you use to discover that certain coefficients were very small but not zero? And how did you force them to be zero? 

I have a very large linear program and manually inspecting values is not feasible.

**Comment by benoit9126** [comment]
Hi @kevinbishop-eaua As far as I remember, I tracked the origin of the problem in a small instance of the problem, thus it was possible to check the values manually. Furthermore, when I exported the problem in a mps file, I realized that there are more zeros than expected (there are some roundings when the data are written in a file).

Now, I use [ortools](https://github.com/google/or-tools) as modeller. You can also use CBC with this modeller.

**Comment by kevinbishop-eaua** [comment]
Thanks for the reply @benoit9126 ! 
I haven't been able to test out your solution yet because the bug is rare and difficult to reproduce. However the next time I run into the problem I'll follow your advice. Will post any findings here in case someone else finds this thread.
Thanks again!

**Comment by benoit9126** [comment]
@waltherg I am now only using `ortools` for modelling linear problems. It is efficient, and I have never faced such issue.

When I was looking for a modeller, I also tried [cvxpy](https://www.cvxpy.org/). It :
* can solve a wider variety of optimization problems (than `ortools`)
* is based on solid mathematical theory (Disciplined Convex Programming)
* is developed by talented people of Stanford University
* is also compatible with CBC (for linear problems)

As far as I remember, it was more time-consuming than `ortools` (this issue is 4 years old, so it may have changed). As my problem was linear, and as I wanted speed (for production), I chose `ortools`. I don't regret it.
If you are not sure what is the best way to model your problem, or if you want to explore options (for research, for instance), `cvxpy` can be very useful to try several modelling options using the same modeller.
